### PR TITLE
add SELECT ALL syntax

### LINKED ISSUE #18706: SQL compatibility: support SELECT ALL syntax
**Use case**
For #15112

**Describe the solution you'd like**

SELECT ALL is identical to SELECT without DISTINCT.

If ALL is specified - ignore it.
If both ALL and DISTINCT are specified - syntax error or exception.

ALL can also be specified inside aggregate function:

SELECT sum(ALL x)

with the same effect (noop).

### PR DESCRIPTION
**Pull request by ucasfl**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add `SELECT ALL` syntax. closes [#18706](https://github.com/ClickHouse/ClickHouse/issues/18706)

### CODE REVIEW

File: src/Parsers/ParserSelectQuery.cpp [hunk 1]
```diff
@@ -93,12 +95,22 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
...
+            has_all = true;
+
         if (s_distinct.ignore(pos, expected))
             select_query->distinct = true;
 
+        if (!has_all && s_all.ignore(pos, expected))
+            has_all = true;
+
+        if (has_all && select_query->distinct)
+            throw Exception("Can not use DISTINCT alongside ALL", ErrorCodes::LOGICAL_ERROR);
```
Comment (alexey-milovidov) [inline src/Parsers/ParserSelectQuery.cpp:112]: It should be syntax error, not logical error.

File: tests/queries/0_stateless/01632_select_all_syntax.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+SELECT ALL 'a';
+SELECT DISTINCT 'a';
+SELECT ALL * FROM (SELECT 1 UNION ALL SELECT 1);
+SELECT DISTINCT * FROM (SELECT 2 UNION ALL SELECT 2);
+SELECT ALL DISTINCT 1; -- { clientError 49 }
+SELECT DISTINCT ALL 1; -- { clientError 49 }
+
+SELECT sum(number) FROM numbers(10);
+SELECT sum(ALL number) FROM numbers(10);
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01632_select_all_syntax.sql:9]: Need to also test that
```
sum(ALL DISTINCT number)
sum(DISTINCT ALL number)
```
yields an error.

File: src/Parsers/ParserSelectQuery.cpp (hunk 1 above)
Comment (ucasFL) [reply src/Parsers/ParserSelectQuery.cpp]: Ok.

**Comment by alexey-milovidov** [comment]
One more detail: we should not throw exceptions from Parser (need to just return false instead).
Because Query Fuzzer is incompatible with it.

**Comment by ucasFL** [comment]
> One more detail: we should not throw exceptions from Parser (need to just return false instead).
> Because Query Fuzzer is incompatible with it. Example:
> https://clickhouse-test-reports.s3.yandex.net/16366/6ec602ad67843a3e2554aed4b1f3fefd10d6ec52/fuzzer/fuzzer.log

There are many `throw Exception` in parser, should we remove all ot them?

**Comment by alexey-milovidov** [comment]
> There are many throw Exception in parser, should we remove all of them?

Not necessarily, there are some useful cases... I will fix Fuzzer in another way.

**Comment by alexey-milovidov** [comment]
Now we have to fix the remaining cases:

```
f(ALL)
f(ALL, x)
```
ALL should be treat as identifier.

```
f(ALL x)
```
Here ALL should be treat as noop.

**Comment by alexey-milovidov** [comment]
Maybe also do the same fixes for DISTINCT:

```
SELECT f(DISTINCT)
SELECT f(DISTINCT, x)
```

**Comment by ucasFL** [comment]
> Now we have to fix the remaining cases:
> 
> ```
> f(ALL)
> f(ALL, x)
> ```
> 
> ALL should be treat as identifier.
> 
> ```
> f(ALL x)
> ```
> 
> Here ALL should be treat as noop.

But if the query is `SELECT f(ALL) FROM table`, what does it mean?

**Comment by alexey-milovidov** [comment]
It means that the table presumable has a column named `ALL`.
Example:

```
SELECT exp(ALL) FROM (SELECT 1 AS ALL)
```

**Comment by alexey-milovidov** [comment]
#19029

**Comment by alexey-milovidov** [comment]
I cannot understand why "Docs check" has failed. Just removed the link.

**Pull request merged**